packages/connect-node/src/grpc-web-handler.ts

```diff
@@ -0,0 +1,182 @@
+      supportedContentType: contentTypeRegExp,
+      requestPath: createMethodUrl("/", spec.service, spec.method),
+      service: spec.service,
+      method: spec.method,
+    });
+  }
+
+  fact.protocolName = protocolName;
+  return fact;
+}
+
+function createHandler<I extends Message<I>, O extends Message<O>>(
+  opt: ProtocolHandlerFactInit,
+  trailerSerialization: Serialization<Headers>,
+  spec: ImplSpec<I, O>
+) {
+  const serialization = createMethodSerializationLookup(
+    spec.method,
+    opt.binaryOptions,
+    opt.jsonOptions
+  );
+  return function handle(req: UniversalRequest): UniversalResponse {
+    const type = parseContentType(req.header.get(headerContentType));
+    if (type == undefined || type.text) {
+      return uResponseUnsupportedMediaType;
+    }
+    if (req.method !== methodPost) {
+      return uResponseMethodNotAllowed;
+    }
+    const context = createHandlerContext(
+      spec,
+      req.header,
+      {
+        [headerContentType]: type.binary ? contentTypeProto : contentTypeJson,
+      },
+      {
+        [headerGrpcStatus]: grpcStatusOk,
+      }
+    );
+    const compression = compressionNegotiate(
+      opt.acceptCompression,
+      req.header.get(headerEncoding),
+      req.header.get(headerAcceptEncoding),
+      headerAcceptEncoding
+    );
+    if (compression.response) {
+      context.responseHeader.set(headerEncoding, compression.response.name);
+    }
+    const outputIt = pipe(
+      req.body,
+      transformPrepend<Uint8Array>(() => {
+        // raise compression error to serialize it as a trailer status
+        if (compression.error) throw compression.error;
+        return undefined;
+      }),
+      transformSplitEnvelope(opt.readMaxBytes),
+      transformDecompressEnvelope(compression.request, opt.readMaxBytes),
+      transformParseEnvelope(
+        serialization.getI(type.binary),
+        trailerFlag,
+        null
+      ),
+      transformInvokeImplementation<I, O>(spec, context),
+      transformSerializeEnvelope(
+        serialization.getO(type.binary),
+        opt.writeMaxBytes
+      ),
+      transformCatchFinally<EnvelopedMessage>((e) => {
+        const trailer = context.responseTrailer;
+        if (e !== undefined) {
+          if (e instanceof ConnectError) {
+            setTrailerStatus(trailer, e);
+          } else {
+            setTrailerStatus(
+              trailer,
+              new ConnectError(
+                "internal error",
+                Code.Internal,
+                undefined,
+                undefined,
+                e
+              )
+            );
+          }
+        }
+        return {
+          flags: trailerFlag,
+          data: trailerSerialization.serialize(trailer),
+        };
+      }),
+      transformCompressEnvelope(compression.response, opt.compressMinBytes),
+      transformJoinEnvelopes()
+    );
+    return {
+      ...uResponseOk,
+      body: outputIt,
+      header: context.responseHeader,
+    };
+  };
```

Comment:
This is the implementation of gRPC-web handlers and shows the meat of this PR. Granted, we are shifting a lot of the complexity into connect-core, but the result seems more manageable than what we previously had.